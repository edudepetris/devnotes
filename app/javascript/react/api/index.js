const getNoteById = (id) => {
  const url = `/api/frontend/v1/notes/${id}`
  const init = {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  }

  return fetch(url, init)
}

/* eslint-disable import/prefer-default-export */
export {getNoteById}
/* eslint-enable import/prefer-default-export */
