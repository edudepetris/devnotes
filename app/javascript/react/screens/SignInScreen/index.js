import {hot} from 'react-hot-loader/root'
import React, {useState} from 'react'
import PropTypes from 'prop-types'

import {useFormik} from 'formik'
import * as Yup from 'yup'

import {
  FormControl,
  FormLabel,
  Alert,
  Icon,
  Box,
  Text,
  Flex,
  Heading,
  Input,
  Button,
  AlertIcon,
  AlertDescription,
  CircularProgress,
  InputGroup,
  InputRightElement,
} from '@chakra-ui/core'
import ThemeToggler from '../../components/ThemeToggler'
import ThemeProvider from '../../components/ThemeProvider'

const LoginForm = ({
  authenticityToken,
  errorMessage,
  resource,
  infoMessage,
  signInPath,
}) => {
  const [isLoading, setIsLoading] = useState(false)
  const [showPassword, setShowPassword] = useState(false)
  const handlePasswordVisibility = () => setShowPassword(!showPassword)

  /* eslint-disable react/forbid-prop-types */
  const {
    touched,
    errors,
    values,
    handleChange,
    handleBlur,
    handleSubmit,
  } = useFormik({
    initialValues: {
      user: {
        email: (resource && resource.email) || '',
        password: '',
      },
    },
    validationSchema: Yup.object().shape({
      user: Yup.object({
        email: Yup.string().email('Invalid email address').required('Required'),
        password: Yup.string().required('Required'),
      }),
    }),
    onSubmit: () => {
      setIsLoading(true)
      document.querySelector('form').submit()
    },
  })
  /* eslint-enable react/forbid-prop-types */

  return (
    <Box>
      <Box textAlign="center">
        <Heading>Login</Heading>
      </Box>
      <Box my={4} textAlign="left">
        <form
          method="POST"
          acceptCharset="UTF-8"
          action={signInPath}
          onSubmit={handleSubmit}
        >
          <input
            type="hidden"
            name="authenticity_token"
            value={authenticityToken}
          />
          <input type="hidden" name="commit" value="Login" />

          {errorMessage && (
            <Box my={4}>
              <Alert status="error" borderRadius={4}>
                <AlertIcon />
                <AlertDescription>{errorMessage}</AlertDescription>
              </Alert>
            </Box>
          )}

          {infoMessage && (
            <Box my={4}>
              <Alert status="info" borderRadius={4}>
                <AlertIcon />
                <AlertDescription>{infoMessage}</AlertDescription>
              </Alert>
            </Box>
          )}

          <FormControl>
            <FormLabel>Email</FormLabel>
            <Input
              type="email"
              placeholder="test@test.com"
              boxSize="lg"
              name="[user][email]"
              id="user_email"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.user.email}
            />
            {touched.user &&
            touched.user.email &&
            errors.user &&
            errors.user.email ? (
              <Text mt={1} variant="caption" color="red.500">
                {errors.user.email}
              </Text>
            ) : null}
          </FormControl>

          <FormControl mt={6}>
            <FormLabel>Password</FormLabel>
            <InputGroup>
              <Input
                type={showPassword ? 'text' : 'password'}
                placeholder="*******"
                boxSize="lg"
                name="[user][password]"
                id="user_password"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.user.password}
              />
              <InputRightElement width="3rem">
                <Button
                  h="1.5rem"
                  boxSize="sm"
                  onClick={handlePasswordVisibility}
                  data-testid="password-view"
                >
                  {showPassword ? (
                    <Icon name="view-off" />
                  ) : (
                    <Icon name="view" />
                  )}
                </Button>
              </InputRightElement>
            </InputGroup>
            {touched.user &&
            touched.user.password &&
            errors.user &&
            errors.user.password ? (
              <Text mt={1} variant="caption" color="red.500">
                {errors.user.password}
              </Text>
            ) : null}
          </FormControl>
          <Button
            colorScheme="teal"
            variant="outline"
            type="submit"
            width="full"
            mt={4}
          >
            {isLoading ? (
              <CircularProgress isIndeterminate w="24px" h="24px" color="teal" />
            ) : (
              'Sign In'
            )}
          </Button>
        </form>
      </Box>
    </Box>
  )
}

LoginForm.propTypes = {
  resource: PropTypes.shape({
    email: PropTypes.string,
  }).isRequired,
  authenticityToken: PropTypes.string.isRequired,
  errorMessage: PropTypes.string.isRequired,
  infoMessage: PropTypes.string.isRequired,
  signInPath: PropTypes.string.isRequired,
}

const SignInScreen = ({
  authenticityToken,
  errorMessage,
  resource,
  infoMessage,
  signInPath,
  signUpPath,
  resetPasswordPath,
}) => {
  return (
    <ThemeProvider>
      <ThemeToggler />
      <Flex width="full" align="center" justifyContent="center">
        <Box
          p={8}
          maxWidth="500px"
          borderWidth={1}
          borderRadius={8}
          boxShadow="lg"
        >
          <LoginForm
            authenticityToken={authenticityToken}
            errorMessage={errorMessage}
            resource={resource}
            infoMessage={infoMessage}
            signInPath={signInPath}
          />
          <Flex justify="space-between">
            <Text as="a" href={signUpPath} color="teal.400" fontSize="sm">
              Sign up
            </Text>
            <Text
              as="a"
              href={resetPasswordPath}
              color="teal.400"
              fontSize="sm"
            >
              Forgot your password?
            </Text>
          </Flex>
        </Box>
      </Flex>
    </ThemeProvider>
  )
}

SignInScreen.propTypes = {
  resource: PropTypes.shape({
    email: PropTypes.string,
  }).isRequired,
  authenticityToken: PropTypes.string.isRequired,
  errorMessage: PropTypes.string.isRequired,
  infoMessage: PropTypes.string.isRequired,
  signInPath: PropTypes.string.isRequired,
  signUpPath: PropTypes.string.isRequired,
  resetPasswordPath: PropTypes.string.isRequired,
}

export default hot(SignInScreen)
