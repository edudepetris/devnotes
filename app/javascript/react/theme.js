// https://chakra-ui.com/docs/theming/customize-theme
import React from 'react' // eslint-disable-line no-unused-vars
import {extendTheme} from '@chakra-ui/react'

const config = {
  initialColorMode: 'dark',
  useSystemColorMode: true,
}

const devnotesTheme = extendTheme({config})

export default devnotesTheme
