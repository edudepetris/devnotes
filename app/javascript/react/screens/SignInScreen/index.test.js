import React from 'react'
import {render, fireEvent, waitFor, screen} from '@testing-library/react'
import withAllTheProviders from '../../test_helpers'
import SignInScreen from './index'

describe('SignInScreen', () => {
  it('renders actions links', () => {
    const {getByText} = render(
      withAllTheProviders(
        <SignInScreen
          authenticityToken="test"
          errorMessage=""
          resource={{email: ''}}
          infoMessage=""
          signInPath="test"
          signUpPath="test"
          resetPasswordPath="test"
        />,
      ),
    )

    expect(getByText('Sign up').href).toBe('http://localhost/test')
    expect(getByText('Forgot your password?').href).toBe(
      'http://localhost/test',
    )
  })

  it('shows/hide password', () => {
    const {getByPlaceholderText, getByTestId, asFragment} = render(
      withAllTheProviders(
        <SignInScreen
          authenticityToken="test"
          errorMessage="test"
          resource={{email: ''}}
          infoMessage="test"
          signInPath="test"
          signUpPath="test"
          resetPasswordPath="test"
        />,
      ),
    )

    fireEvent.change(getByPlaceholderText('*******'), {
      target: {value: 'secret'},
    })
    const firstRender = asFragment()

    fireEvent.click(getByTestId('password-view'))

    expect(firstRender).toMatchDiffSnapshot(asFragment())
  })

  it('requires email and password', async () => {
    const {getByText, getAllByText} = render(
      withAllTheProviders(
        <SignInScreen
          authenticityToken="test"
          errorMessage="test"
          resource={{email: ''}}
          infoMessage="test"
          signInPath="test"
          signUpPath="test"
          resetPasswordPath="test"
        />,
      ),
    )

    await waitFor(() => {
      fireEvent.click(getByText('Sign In'))
      expect(getAllByText('Required').length).toBe(2)
    })
  })

  it('renders info message', () => {
    render(
      withAllTheProviders(
        <SignInScreen
          authenticityToken="test"
          errorMessage=""
          resource={{email: ''}}
          infoMessage="info message"
          signInPath="test"
          signUpPath="test"
          resetPasswordPath="test"
        />,
      ),
    )

    expect(screen.getByText('info message')).toBeTruthy()
  })

  it('renders error message', () => {
    render(
      withAllTheProviders(
        <SignInScreen
          authenticityToken="test"
          errorMessage="error message"
          resource={{email: ''}}
          infoMessage=""
          signInPath="test"
          signUpPath="test"
          resetPasswordPath="test"
        />,
      ),
    )

    expect(screen.getByText('error message')).toBeTruthy()
  })
})
