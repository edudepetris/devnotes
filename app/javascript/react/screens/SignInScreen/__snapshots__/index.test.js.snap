// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInScreen shows/hide password 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -236,11 +236,11 @@
                  <input
                    class=\\"chakra-input css-17hoxe6\\"
                    id=\\"user_password\\"
                    name=\\"[user][password]\\"
                    placeholder=\\"*******\\"
-                   type=\\"password\\"
+                   type=\\"text\\"
                    value=\\"secret\\"
                  />
                  <div
                    class=\\"chakra-input__right-element css-il15dq\\"
                  >
@@ -256,16 +256,14 @@
                      >
                        <g
                          fill=\\"currentColor\\"
                        >
                          <path
-                           d=\\"M23.432,10.524C20.787,7.614,16.4,4.538,12,4.6,7.6,4.537,3.213,7.615.568,10.524a2.211,2.211,0,0,0,0,2.948C3.182,16.351,7.507,19.4,11.839,19.4h.308c4.347,0,8.671-3.049,11.288-5.929A2.21,2.21,0,0,0,23.432,10.524ZM7.4,12A4.6,4.6,0,1,1,12,16.6,4.6,4.6,0,0,1,7.4,12Z\\"
+                           d=\\"M23.2,10.549a20.954,20.954,0,0,0-4.3-3.6l4-3.995a1,1,0,1,0-1.414-1.414l-.018.018a.737.737,0,0,1-.173.291l-19.5,19.5c-.008.007-.018.009-.026.017a1,1,0,0,0,1.631,1.088l4.146-4.146a11.26,11.26,0,0,0,4.31.939h.3c4.256,0,8.489-2.984,11.051-5.8A2.171,2.171,0,0,0,23.2,10.549ZM16.313,13.27a4.581,4.581,0,0,1-3,3.028,4.3,4.3,0,0,1-3.1-.19.253.253,0,0,1-.068-.407l5.56-5.559a.252.252,0,0,1,.407.067A4.3,4.3,0,0,1,16.313,13.27Z\\"
                          />
-                         <circle
-                           cx=\\"12\\"
-                           cy=\\"12\\"
-                           r=\\"2\\"
+                         <path
+                           d=\\"M7.615,13.4a.244.244,0,0,0,.061-.24A4.315,4.315,0,0,1,7.5,12,4.5,4.5,0,0,1,12,7.5a4.276,4.276,0,0,1,1.16.173.244.244,0,0,0,.24-.062l1.941-1.942a.254.254,0,0,0-.1-.421A10.413,10.413,0,0,0,12,4.75C7.7,4.692,3.4,7.7.813,10.549a2.15,2.15,0,0,0-.007,2.9,21.209,21.209,0,0,0,3.438,3.03.256.256,0,0,0,.326-.029Z\\"
                          />
                        </g>
                      </svg>
                    </button>
                  </div>"
`;
